### PR TITLE
feat (customization): #2275 top sites drag and drop

### LINKED ISSUE #2275: Edit Top Sites: implement drag and drop
From #1987:  UI for the panel allows for drag and drop of Top Site cards

This one depends on pinning #2274

### PR DESCRIPTION
**Pull request by rlr**
Fixes #2275

### CODE REVIEW

File: addon/Feeds/TopSitesFeed.js [hunk 1]
```diff
@@ -49,7 +49,27 @@ module.exports = class TopSitesFeed extends Feed {
...
+
+    // Return the new set of top sites.
+    this._getLinks()
+      .then(links => this.options.send(am.actions.Response("TOP_FRECENT_SITES_RESPONSE", links), action.workerId, true))
+      .catch(Cu.reportError);
+  }
   _insertPin(site, index) {
+    // Don't insert any pins past the end of the visible top sites. Otherwise,
+    // we can end up with a bunch of pinned sites that can never be unpinned again
+    // from teh UI.
```
Comment (ncloudioj) [inline addon/Feeds/TopSitesFeed.js]: typo: s/teh/the/

File: content-test/addon/Feeds/TopSitesFeed.test.js [hunk 2]
```diff
@@ -250,4 +259,31 @@ describe("TopSitesFeed", () => {
...
+    it("should move a pinned site in specified slot to the next slot", () => {
+      const site1 = {url: "example.com"};
+      instance.pinnedLinks.links = [null, null, site1];
+      const site = {url: "foo.bar", title: "foo"};
+      instance.dropTopSite({data: {index: 2, url: site.url, title: site.title}});
+      assert.calledTwice(instance.pinnedLinks.pin);
+      assert.calledWith(instance.pinnedLinks.pin, site, 2);
+      assert.calledWith(instance.pinnedLinks.pin, site1, 3);
+    });
+    it("should move not move a pinned site in specified slot past the end of the list", () => {
```
Comment (ncloudioj) [inline content-test/addon/Feeds/TopSitesFeed.test.js]: ditto: s/should move not/should not/

File: addon/Feeds/TopSitesFeed.js [hunk 3]
```diff
@@ -49,7 +49,27 @@ module.exports = class TopSitesFeed extends Feed {
...
+    if (index >= TOP_SITES_SHOWMORE_LENGTH) {
```
Comment (ncloudioj) [inline addon/Feeds/TopSitesFeed.js]: A nice guard here!

**Comment by as-pine-proxy** [comment]
[Treeherder: 18d1c34340ce39cebc23b6b8f1c93d5634c19846](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=18d1c34340ce39cebc23b6b8f1c93d5634c19846)

**Pull request merged**